src/main/java/com/sourcegraph/cody/agent/CodyAgentClient.java

```diff
@@ -26,14 +33,112 @@ public class CodyAgentClient {
   // onSetConfigFeatures
   @Nullable public Consumer<WebviewPostMessageParams> onReceivedWebviewMessage;
 
-  @Nullable public Editor editor;
+  // Callback for the "editTask/didUpdate" notification from the agent.
+  @Nullable private Consumer<EditTask> onEditTaskDidUpdate;
+
+  // Callback for the "editTask/didDelete" notification from the agent.
+  @Nullable private Consumer<EditTask> onEditTaskDidDelete;
+
+  // Callback for the "textDocument/edit" request from the agent.
+  @Nullable private Consumer<TextDocumentEditParams> onTextDocumentEdit;
+
+  // Callback for the "workspace/executeCommand" request from the agent.
+  @Nullable private Consumer<DisplayCodeLensParams> onDisplayCodeLens;
+
+  // Callback for the "workspace/edit" request from the agent.
+  @Nullable private Consumer<WorkspaceEditParams> onWorkspaceEdit;
+
+  public void setOnEditTaskDidUpdate(@Nullable Consumer<EditTask> callback) {
+    onEditTaskDidUpdate = callback;
+  }
+
+  public void setOnEditTaskDidDelete(@Nullable Consumer<EditTask> callback) {
+    onEditTaskDidDelete = callback;
+  }
+
+  @JsonNotification("editTask/didUpdate")
+  public void editTaskDidUpdate(EditTask params) {
+    onEventThread(
+        () -> {
+          if (onEditTaskDidUpdate != null) {
+            onEditTaskDidUpdate.accept(params);
+          } else {
+            logger.warn("No callback registered for editTask/didUpdate");
+          }
+          return null;
+        });
+  }
+
+  @JsonNotification("editTask/didDelete")
+  public void editTaskDidDelete(EditTask params) {
+    onEventThread(
+        () -> {
+          if (onEditTaskDidDelete != null) {
+            onEditTaskDidDelete.accept(params);
+          } else {
+            logger.warn("No callback registered for editTask/didDelete");
+          }
+          return null;
+        });
+  }
+
+  public void setOnTextDocumentEdit(@Nullable Consumer<TextDocumentEditParams> callback) {
+    onTextDocumentEdit = callback;
+  }
+
+  @JsonRequest("textDocument/edit")
+  public CompletableFuture<Boolean> textDocumentEdit(TextDocumentEditParams params) {
+    return onEventThread(
+        () -> {
+          if (onTextDocumentEdit != null) {
+            onTextDocumentEdit.accept(params);
+          } else {
+            logger.warn("No callback registered for textDocument/edit");
+          }
+          return true;
```

Comment:
Unrelated to this change, but I think a lot of our future hopes and dreams hinge on this boolean.

The VSCode edit API can fail, and this boolean indicates the success or failure of the edit. When we have concurrent edits (in the sense that the Agent's model of the document isn't caught up to the JetBrains reality because the user has edited but the extension sends an edit), we could detect that and fail here.

Punchline is that extension code doesn't handle those failures! We would need to have the extension handle that. But there's probably some *not terribly burdensome* abstraction, a "retryable edit", that could do that. Interested in your thoughts here.

Reply:
Yes, let's do it -- I've seen it fail to apply once (no code returned by LLM) and would love to get more feedback to the user when this happens.